File: Binary Tree Java/Binary Tree Preorder Traversal/IterativeSolution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    static List<Integer> res;
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res= new ArrayList<>();
        if(root==null)
            return res;
        
        Stack<TreeNode> S= new Stack<>();
        S.add(root);
        while(!S.isEmpty()) {
            TreeNode temp=S.pop();
            
            res.add(temp.val);
            
            //Add right then left
            //So that left occurs on top when poping
            if(temp.right!=null)
                S.add(temp.right);
            
            if(temp.left!=null)
                S.add(temp.left);
        }
        
        return res;        
    }
    
}

File: Binary Tree Java/Binary Tree Preorder Traversal/IterativeSolution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun preorderTraversal(root: TreeNode?): List<Int> {   
        val res= ArrayList<Int>()
        
        root?: return res
        val temp = Stack<TreeNode>()
        temp.add(root)
        while(!temp.isEmpty()) {
            val poped = temp.pop()
            
            res.add(poped.`val`)
            if(poped.right!=null)
                temp.add(poped.right)
                
            if(poped.left!=null)
                temp.add(poped.left)
        }
        
        return res
    }
}
